### ISSUE TITLE
Value cannot be null exception in ExpressiveValidator GetBasicRule

### ISSUE THREAD

**Issue by programcsharp** [JaroslawWaliszko/ExpressiveAnnotations#80]
I'm getting error reports from users that are strange. Looks like an issue with cached expression validators, but I haven't been able to find repro steps. If a user finds the issue, I can repro it, but I can't figure out how to repro it from scratch. Once I reset the app pool, everything works fine again until the next time the issue occurs, including refreshing that page a few times.

Here's a snapshot of the stack trace the user sees:
![unnamed](https://cloud.githubusercontent.com/assets/994270/10317179/05a0f198-6c18-11e5-84ba-461b8edb4a9e.png)

Looks like the issue occurs here: https://github.com/JaroslawWaliszko/ExpressiveAnnotations/blob/master/src/ExpressiveAnnotations.MvcUnobtrusive/Validators/ExpressiveValidator.cs#L139 -- should those 3 .Any() calls be null checked (or not happen at all) if Cached == true? Seems like that would solve the issue, but I didn't want to PR without knowing more about the caching and what should be going on here.

**Comment by JaroslawWaliszko** [comment]
I've pushed a suggested change related to this issue: https://github.com/JaroslawWaliszko/ExpressiveAnnotations/commit/d147abfcbf730e5a0e5cf70b0242df739aa9668b. Nevertheless I consider this exception very unlikely, since I cannot think of any circumstances, under which it can possibly happen. This maps (`FieldsMap`, etc.) should be always **non-null** there. If anything, they should be empty, not null (the constructor takes care of that). 

Nevertheless these checks can work here as an additional level of safety from the pragmatic point of view. 

Finally, because I think this issue should have never happened, I've put an assertion check there. If assembly is compiled in the debug mode, it will notify developers if the case occurs (in release mode nothing will happen).

**Comment by programcsharp** [comment]
Yes, that's what I thought -- this should never happen. Reading through the code I don't see a way. But somehow it is happening. Those crazy users ;-)

Thanks for the fix! I'll drop it in and see if the problem goes away. I wish I had repro steps so I could prove a fix, but I spent an hour or two and couldn't make it happen.

**Comment by JaroslawWaliszko** [comment]
You're welcome. Also thanks for registration of such an issue. I've divided mentioned assertion into separate lines: https://github.com/JaroslawWaliszko/ExpressiveAnnotations/commit/496e415f1b4c3b06b9434c6f75ecd25ac9dc7643. Maybe it will help to uncover some hidden problem. Regards.

**Comment by JaroslawWaliszko** [comment]
Just one more thing. I've just noticed that the exception message says: 
> ArgumetNullException: value cannot be null. Parameter name: source

It's rather unlikely for these lines to throw that. I've searched the solution and found that I'm throwing such an exception from the inside of `ExtractValue` method, defined in `Helper` type:

```
public static object ExtractValue(object source, string property)
{
    if (source == null)
        throw new ArgumentNullException("source");
```

This one in turn, is only called from inside of `ExtractValuesForMessage` method, defined in `ExpressiveAttribute` type. While the call stack does not point at that line, just out of curiosity, are the users using curly brackets `{` somewhere in their error messages? I'm asking because from [2.6.2 version](https://github.com/JaroslawWaliszko/ExpressiveAnnotations/releases/tag/v2.6.2) and above there is an enhancement which tries to extract model values from fields, which names are placed inside such brackets.

**Comment by programcsharp** [comment]
I think the exception is coming from .NET framework code inside the .Any() call. If I disassemble that, it does:

if (source == null)
    throw Error.ArgumentNull("source");

**Comment by programcsharp** [comment]
I did have a question on the versioning, though. We've got the latest nuget (ExpressiveAnnotations 2.6.3). But that has ExpressiveAnnotations.dll v 2.3.3 and ExpressiveAnnotations.dll v 2.4.2. Are those versions right? And why are they all different version numbers?

**Comment by JaroslawWaliszko** [comment]
These version numbers are correct (I realize this can be initially confusing though).

They are all different, because e.g. 2.6.3 version of ExpressiveAnnotations NuGet package, is an overall product version - product which contains of 3 separate components: ExpressiveAnnotations.dll, ExpressiveAnnotations.MvcUnobtrusive.dll and expressive.annotations.validate.js. Each of these componentt is independently versioned. 

Because many NuGet packages contain only single component, there is no discrepancy between the package version and such an component (assembly/script/etc.) version.

At the initial stage of working at this library I had been thinking of how to version the releases. At first I thought the highest version of all the components will automatically define the version of the package. But after a while I realized it is not always possible, and manual versioning is needed - example is below:
```
initial version ---> A.dll v1.0, B.dll v1.0, C.js v1.0 ---> release ---> NuGet package v1.0
bug fix for A   ---> A.dll v1.1, B.dll v1.0, C.js v1.0 ---> release ---> NuGet package v1.1
bug fix for B   ---> A.dll v1.1, B.dll v1.1, C.js v1.0 ---> release ---> NuGet package v?.? 
                                                                              (probably 1.2)
bug fix for B   ---> A.dll v1.1, B.dll v1.2, C.js v1.0 ---> release ---> NuGet package v?.? 
                                                                           (most likely 1.3)
feature for C   ---> A.dll v1.1, B.dll v1.2, C.js v2.0 ---> release ---> NuGet package v?.? 
                                                                           (most likely 2.0)
```
Regards

**Comment by theyetiman** [comment]
Just to add my 2 pennies worth... I've received the exact same error today (odd coincidence?). I recycled the apppool and the problem was fixed. This leads me to believe it's something to do with the state of the cached FieldsMap/ConstsMap/ParsersMap but like @programcsharp I don't know enough about what's going on to work it out fully.

**Comment by JaroslawWaliszko** [comment]
Maybe there was a coincidence related to single item expiration. Here: https://github.com/JaroslawWaliszko/ExpressiveAnnotations/commit/f0a75543856571d49a2840761f72e3689672cad7, I've changed 
```
private bool Cached
{
    get { return FieldsMap != null || ConstsMap != null || ParsersMap != null; }
```
into 
```
private bool Cached
{
    get { return FieldsMap != null && ConstsMap != null && ParsersMap != null; }
```
Previously, if any map was positively retrieved from cache, the constructor logic was sure that other maps were also there (non-null). But maybe there are cases where some single item is removed from the cache, while the others are still there (earlier expiration? removal from outside component?). Because of that, now we're checking for existence of all 3 items (`FieldsMap`, `ConstsMap` and `ParsersMap`). If any is missing, parsing is reinvoked.

**Comment by theyetiman** [comment]
This is great @JaroslawWaliszko. Cheers!

**Comment by programcsharp** [comment]
Ahh, that's a good point. Brings up a couple questions:
 - Is there a reason to use HttpRuntime.Cache vs. a static dictionary? I suppose if there is a lot of memory usage stuff can be nudged out, but I don't think that's the case here?
 - Why is the lock for? _locker isn't static, so it's only per instance. Also, you have a double checked lock:

``` c#
                if (!Cached) // is the cache empty?
                {
                    lock (_locker)
                    {
                        if (!Cached)
                        {
```

But you only load from cache once above, so it could never change?

**Comment by theyetiman** [comment]
@programcsharp I can't say for sure that I know it's necessary here, but the double check you mention is a common pattern to ensure the code (intended to run only once) inside the `lock` statement doesn't run multiple times when multiple threads are waiting for the lock. Say that did happen, and no second check were done, then the current thread would wait for the lock, get it, then do the same processing again that the other thread (which previously had the lock) has already done.

I don't know enough about `DataAnnotationsModelValidator` to know how these classes really work but at a glance it does look like the lock should be a static var. Maybe @JaroslawWaliszko can clear it up.

**Comment by programcsharp** [comment]
Sure, but my point is for the double checked lock to be useful, you've got to load again. And the _locker has to be static. In other words:

``` c#
        private static readonly object _locker = new object();


                FieldsMap = HttpRuntime.Cache.Get(fieldsId) as IDictionary<string, string>;
                ConstsMap = HttpRuntime.Cache.Get(constsId) as IDictionary<string, object>;
                ParsersMap = HttpRuntime.Cache.Get(parsersId) as IDictionary<string, string>;

                if (!Cached) // is the cache empty?
                {
                    lock (_locker)
                    {
                        // *check again inside lock*
                        FieldsMap = HttpRuntime.Cache.Get(fieldsId) as IDictionary<string, string>;
                        ConstsMap = HttpRuntime.Cache.Get(constsId) as IDictionary<string, object>;
                        ParsersMap = HttpRuntime.Cache.Get(parsersId) as IDictionary<string, string>;

                        if (!Cached)
                        {
```

**Comment by theyetiman** [comment]
Ignore my response... I didn't see your last paragraph. Last thing I read was "Also, you have a double checked lock:" so I was just trying to explain the pattern rather than the implementation but it's clear you already understand.

**Comment by programcsharp** [comment]
PR #82 is an example of what I'm talking about to remove the HttpCache issue and the lock.

**Comment by JaroslawWaliszko** [comment]
The threading-related logic was added due to this issue: https://github.com/JaroslawWaliszko/ExpressiveAnnotations/issues/34. 

Previously there was no locking there because I've seen it unnecessary, as no multiple threads should be there in the constructor. Unfortunately I wasn't been able to reproduce that problem (and fix it iteratively), since I've never been using this lib in multi-threaded selenium testing environment. So it basically have been added there just in case of. 

To be honest I still see this part as something redundant. If only I'd be able to reproduce issue from referenced issue above, I'd be able to explain or remove this.

Instead of framework provided Cache, the static dictionary should be good here as well, since it should be there as long as possible (until the pool recycles). The caching is here because each request seems to create new Validator objects, but because parsing is expensive and metadata unchanged, it's good to have it all processed asap.

The double check lock has been used due to some possible issues with memory model / compiler optimizations. Because I've been thinking about that some time ago I'll just provide the link to my thinking: http://stackoverflow.com/a/13262058/270315. Secondary cache accessing wasn't the case here.

**Comment by programcsharp** [comment]
Right -- that's why i based #82 on ConcurrentDictionary. That should retain all the multithreading support while also still caching, but doing it on a static ConcurrentDictionary rather than HttpCache.

**Comment by pawelpabich** [comment]
The only thing I can add is that we have not seen the problem since the lock was introduced.  I know this does not explain the "why part" but that's all I have right now.

**Comment by jwaliszko** [comment]
Just update: Caching implementation slightly improved. `ConcurrentDictionary` stays as the underlying storage, but value access is modified and guarantees atomic invocation of the delegate value factory. Details and comments here: https://github.com/jwaliszko/ExpressiveAnnotations/commit/8c3cd58090b0025941f428fb6edc74ee01f89018.
